composer/optim/scheduler.py

```diff
@@ -1,226 +1,265 @@
+                                     ssr: float = 1.0,
+                                     warmup_time: Union[str, Time],
+                                     milestones: List[Union[str, Time]],
+                                     gamma: float = 0.1) -> float:
+    warmup_time = _convert_time(warmup_time, state)
+    if state.timer < warmup_time:
+        return linear_scheduler(state, start_factor=0.0, end_factor=1.0, total_time=warmup_time)
+
+    return multi_step_scheduler(state, ssr=ssr, milestones=milestones, gamma=gamma)
+
+
+def linear_with_warmup_scheduler(state: State,
+                                 *,
+                                 ssr: float = 1.0,
+                                 warmup_time: Union[str, Time],
+                                 start_factor: float = 1.0,
+                                 end_factor: float = 0.0,
+                                 total_time: Union[str, Time] = '1dur'):
+    # N.B. warmup time is intentionally *not* subject to scale schedule
+    warmup_time = _convert_time(warmup_time, state)
+    if state.timer < warmup_time:
+        return linear_scheduler(state, start_factor=0.0, end_factor=start_factor, total_time=warmup_time)
+
+    total_time = _convert_time(total_time, state, ssr=ssr)
+    current_time = state.timer.get(warmup_time.unit)
+    frac_of_total = min(1.0, ((current_time - warmup_time) / (total_time - warmup_time)).value)
+
+    current_factor = start_factor + frac_of_total * (end_factor - start_factor)
+
+    return current_factor
 
-        self.optimizer = optimizer
-        self.T_max = T_max
-        self.power = power
-        self.eta_min = eta_min
-        super(PolynomialLR, self).__init__(optimizer, last_epoch, verbose)  # type: ignore
 
-    def get_lr(self):
-        coeff = (1 - self.last_epoch / self.T_max)**self.power  # type: ignore
-        return [(base_lr - self.eta_min) * coeff + self.eta_min for base_lr in self.base_lrs]  # type: ignore
+def cosine_annealing_with_warmup_scheduler(state: State,
+                                           *,
+                                           ssr: float = 1.0,
+                                           warmup_time: Union[str, Time],
+                                           T_max: Union[str, Time] = '1dur',
+                                           min_factor: float = 0.0):
+    # N.B. warmup time is intentionally *not* subject to scale schedule
+    warmup_time = _convert_time(warmup_time, state)
+    if state.timer < warmup_time:
+        return linear_scheduler(state, start_factor=0.0, end_factor=1.0, total_time=warmup_time)
+
+    T_max = _convert_time(T_max, state, ssr=ssr)
+    current_time = state.timer.get(warmup_time.unit)
+    frac_of_total = ((current_time - warmup_time) / (T_max - warmup_time)).value
+
+    return _cosine_anneal(x=frac_of_total, min_y=min_factor)
+
+
+def get_scheduler(scheduler: ComposerSchedulerFn, ssr: float = 1.0, **kwargs) -> ComposerSchedulerFn:
```

Comment:
Maybe rename to `get_scheduler_with_ssr` to be more clear about what it's used for?

Reply:
Woops, this function doesn't need to be here at all actually.